heat/core/linalg/lanczos.py

```diff
@@ -0,0 +1,234 @@
+
+    """
+    if not isinstance(A, dndarray.DNDarray):
+        raise TypeError("A needs to be of type ht.dndarra, but was {}".format(type(A)))
+
+    if not (A.numdims == 2):
+        raise RuntimeError("A needs to be a 2D matrix")
+    if not isinstance(m, (int, float)):
+        raise TypeError("m must be eiter int or float, but was {}".format(type(m)))
+
+    n, column = A.shape
+    if n != column:
+        raise TypeError("Input Matrix A needs to be symmetric.")
+    T = factories.zeros((m, m))
+    if A.split == 0:
+        # This is done for better memory access in the reorthogonalization Gram-Schmidt algorithm
+        V = factories.ones((m, n), split=1, dtype=A.dtype, device=A.device)
+    else:
+        V = factories.ones((m, n), split=None, dtype=A.dtype, device=A.device)
+
+    if v0 is None:
+        vr = random.rand(n)
+        v0 = vr / norm(vr)
+
+    # # 0th iteration
+    # # vector v0 has euklidian norm = 1
+    w = matmul(A, v0)
+    alpha = dot(w, v0)
+    w = w - alpha * v0
+    T[0, 0] = alpha
+    V[0, :] = v0
+    for i in range(1, int(m)):
+        beta = norm(w)
+        if abs(beta) < 1e-10:
+            print("Lanczos breakdown in iteration {}".format(i))
+            # Lanczos Breakdown, pick a random vector to continue
+            vr = random.rand(n, dtype=A.dtype)
+            # orthogonalize v_r with respect to all vectors v[i]
+            for j in range(i):
+                vi_loc = V._DNDarray__array[j, :]
+                a = torch.dot(vr._DNDarray__array, vi_loc)
+                b = torch.dot(vi_loc, vi_loc)
+                A.comm.Allreduce(MPI.IN_PLACE, a, MPI.SUM)
+                A.comm.Allreduce(MPI.IN_PLACE, b, MPI.SUM)
+                vr._DNDarray__array = vr._DNDarray__array - a / b * vi_loc
+            # normalize v_r to Euclidian norm 1 and set as ith vector v
+            vi = vr / norm(vr)
+        else:
+            vr = w
+
+            # Reorthogonalization
+            # ToDo: Rethink this; mask torch calls, See issue #494
+            # This is the fast solution, using item access on the ht.dndarray level is way slower
+            for j in range(i):
+                vi_loc = V._DNDarray__array[j, :]
+                a = torch.dot(vr._DNDarray__array, vi_loc)
+                b = torch.dot(vi_loc, vi_loc)
+                A.comm.Allreduce(MPI.IN_PLACE, a, MPI.SUM)
+                A.comm.Allreduce(MPI.IN_PLACE, b, MPI.SUM)
+                vr._DNDarray__array = vr._DNDarray__array - a / b * vi_loc
+
+            vi = vr / norm(vr)
+
+        w = matmul(A, vi)
+        alpha = dot(w, vi)
+        w = w - alpha * vi - beta * V[i - 1, :]
+
+        T[i - 1, i] = beta
+        T[i, i - 1] = beta
+        T[i, i] = alpha
+        V[i, :] = vi
+
+    if V.split is not None:
+        V.resplit_(axis=None)
+    V = V.transpose()
+    if T_out is not None:
```

Comment:
Inner function assignment is not sufficient if you want to do a copy-over into the tensors.